chatsky/utils/context_dict/ctx_dict.py

```diff
@@ -0,0 +1,203 @@
+        if isinstance(key, slice):
+            for k in range(len(self._keys))[key]:
+                del self[k]
+        else:
+            self._removed += [key]
+            if key in self._items.keys():
+                self._keys.remove(key)
+            if key in self._added:
+                self._added.remove(key)
+            del self._items[key]
+
+    def __iter__(self) -> Sequence[K]:
+        return iter(self._keys if self._storage is not None else self._items.keys())
+    
+    def __len__(self) -> int:
+        return len(self._keys if self._storage is not None else self._items.keys())
+
+    async def get(self, key: K, default: V = _marker) -> V:
+        try:
+            return await self[key]
+        except KeyError:
+            if default is self._marker:
+                raise
+            return default
+
+    def keys(self) -> Set[K]:
+        return set(iter(self))
+
+    def __contains__(self, key: K) -> bool:
+        return key in self.keys()
+
+    async def items(self) -> Set[Tuple[K, V]]:
+        return {(k, await self[k]) for k in self.keys()}
+
+    async def values(self) -> Set[V]:
+        return {await self[k] for k in self.keys()}
+
+    async def pop(self, key: K, default: V = _marker) -> V:
+        try:
+            value = await self[key]
+        except KeyError:
+            if default is self._marker:
+                raise
+            return default
+        else:
+            del self[key]
+            return value
+
+    async def popitem(self) -> Tuple[K, V]:
+        try:
+            key = next(iter(self))
+        except StopIteration:
+            raise KeyError from None
+        value = await self[key]
+        del self[key]
+        return key, value
+
+    async def clear(self) -> None:
+        try:
+            while True:
+                await self.popitem()
+        except KeyError:
+            pass
+
+    async def update(self, other: Any = (), /, **kwds) -> None:
+        if isinstance(other, ContextDict):
+            for key in other:
+                self[key] = await other[key]
+        elif isinstance(other, Mapping):
+            for key in other:
+                self[key] = other[key]
+        elif hasattr(other, "keys"):
+            for key in other.keys():
+                self[key] = other[key]
+        else:
+            for key, value in other:
+                self[key] = value
+        for key, value in kwds.items():
+            self[key] = value
+
+    async def setdefault(self, key: K, default: V = _marker) -> V:
+        try:
+            return await self[key]
+        except KeyError:
+            if default is self._marker:
+                raise
+            self[key] = default
+        return default
+
+    @model_validator(mode="wrap")
+    def _validate_model(value: Dict[K, V], handler: Callable[[Dict], "ContextDict"]) -> "ContextDict":
+        instance = handler(dict())
+        instance._items = {key: value[key] for key in sorted(value)}
```

Comment:
Is there a need to sort keys?

Reply:
It was used to get the last index to know where to insert new elements.
Maybe using `max(keys)` would be better.